File: quarantine_tc.py
#!/usr/bin/env python

# https://gist.github.com/catlee/9f85b4d51425a41cdc33ab8c7b754507/revisions

from __future__ import print_function

import datetime
from taskcluster import Queue
from taskcluster.exceptions import TaskclusterRestFailure
from taskcluster.utils import fromNow
from argparse import ArgumentParser


def main():
    parser = ArgumentParser()
    parser.add_argument('--disable', action='store_true', dest='quarantine',
                        help='disable the workers for 1000 years',
                        default=True)
    parser.add_argument('--enable', action='store_false', dest='quarantine',
                        help='enable the workers')
    parser.add_argument('-p', '--provisioner', required=True)
    parser.add_argument('-w', '--worker-type', required=True)
    parser.add_argument('-g', '--worker-group', required=True)
    parser.add_argument('workers', nargs='+', help='worker ids')

    args = parser.parse_args()

    if args.quarantine:
        quarantineUntil = fromNow('1000 years')
    else:
        quarantineUntil = fromNow('-1 hours')

    q = Queue()
    expire = datetime.datetime.now() + datetime.timedelta(62) # roughly 2 months

    for worker_id in args.workers:
        try:
            res = q.getWorker(args.provisioner, args.worker_type, args.worker_group, worker_id)
        except TaskclusterRestFailure as e:
            if 'ResourceNotFound' in repr(e):
                # Create a dummy if the worker is not in taskcluster
                q.declareWorker(args.provisioner, args.worker_type, args.worker_group, worker_id, {"expires":expire})
            else:
                print('getWorker failed: {}'.format(e))
                continue

        res = q.quarantineWorker(args.provisioner, args.worker_type,
                                 args.worker_group, worker_id,
                                 payload={'quarantineUntil': quarantineUntil })
        if 'quarantineUntil' in res:
            print('{0[workerId]} quarantined until {0[quarantineUntil]}'.format(res))
        else:
            print('{0[workerId]} not quarantined'.format(res))


if __name__ == '__main__':
    main()

File: quarantine_tc.js
#!/usr/bin/env node

// example usage:
// [type] [group] [quarantine_time] [expiration] [hostnames]
// ./quarantine_tc.js gecko-t-linux-talos mdc1 "1 year" "1 day" t-linux64-ms{226..240}
//

var taskcluster = require('taskcluster-client');
taskcluster.config(taskcluster.fromEnvVars());
var q = new taskcluster.Queue({rootUrl:'https://firefox-ci-tc.services.mozilla.com'});
var provisionerId = 'releng-hardware';
var workerType = process.argv[2];
var workerGroup = process.argv[3];
process.argv.slice(6).forEach(function(workerId) {
    console.log(workerId);
    q.getWorker(provisionerId, workerType, workerGroup, workerId)
        .then(function(a) {
            console.log('expires:', a['expires']);
            console.log('quarantine:', a['quarantineUntil']);
            console.log('--');
        }, error => { console.log('caught', error.message); })
        .finally(function(a) {
            if (process.argv[5] != ' ') {
                var exp = taskcluster.fromNow(process.argv[5]);
            } else {
                var exp = taskcluster.fromNow("1 year");
            }
            q.declareWorker(provisionerId, workerType, workerGroup, workerId, {"expires":exp})
                .then(function(a) {
                    q.getWorker(provisionerId, workerType, workerGroup, workerId)
                        .then(function(a){
                            console.log(a['workerId'], 'expires:', a['expires']);
                            if (process.argv[4] != ' ') {
                                var quar = taskcluster.fromNow(process.argv[4]);
                                q.quarantineWorker(provisionerId, workerType, workerGroup, workerId, {"quarantineUntil":quar})
                                    .then(function(a) {
                                        q.getWorker(provisionerId, workerType, workerGroup, workerId)
                                            .then(function(a){
                                                console.log(a['workerId'], 'quarantine:', a['quarantineUntil']);
                                            });
                                    })
                                    .catch(function(z) {
                                        console.warn(z);
                                    });
                            }
                        });
                })
                .catch(function(z) {
                    console.warn(z);
                });
        });
});
